tf2_ros/include/tf2_ros/message_filter.h

```diff
@@ -32,690 +32,550 @@
-      }
-    }
-
-
-    // We can transform already
-    if (info.success_count == expected_success_count_)
-    {
-      messageReady(evt);
-    }
-    else
-    {
-      // If this message is about to push us past our queue size, erase the oldest message
-      if (queue_size_ != 0 && message_count_ + 1 > queue_size_)
-      {
-        // While we're using the reference keep a shared lock on the messages.
-        boost::shared_lock< boost::shared_mutex > shared_lock(messages_mutex_);
 
-        ++dropped_message_count_;
-        const MessageInfo& front = messages_.front();
-        TF2_ROS_MESSAGEFILTER_DEBUG("Removed oldest message because buffer is full, count now %d (frame_id=%s, stamp=%f)", message_count_,
-                                (mt::FrameId<M>::value(*front.event.getMessage())).c_str(), mt::TimeStamp<M>::value(*front.event.getMessage()).toSec());
 
-        V_TransformableRequestHandle::const_iterator it = front.handles.begin();
-        V_TransformableRequestHandle::const_iterator end = front.handles.end();
-        for (; it != end; ++it)
-        {
-          bc_.cancelTransformableRequest(*it);
+        // We can transform already
+        if (info.success_count == expected_success_count_) {
+          messageReady(evt);
         }
+        else {
+          // If this message is about to push us past our queue size, erase the oldest message
+          if (queue_size_ != 0 && message_count_ + 1 > queue_size_) {
+
+            // While we're using the reference keep a shared lock on the messages.
+            std::lock_guard< std::mutex > guard_lock(messages_mutex_);
+
+            ++dropped_message_count_;
+            const MessageInfo& front = messages_.front();
+            TF2_ROS_MESSAGEFILTER_DEBUG("Removed oldest message because buffer is full, count now %d (frame_id=%s, stamp=%f)",
+                                        message_count_,
+                                        (mt::FrameId<M>::value(*front.event.getMessage())).c_str(),
+                                        mt::TimeStamp<M>::value(*front.event.getMessage()).seconds());
+
+            V_TransformableRequestHandle::const_iterator it = front.handles.begin();
+            V_TransformableRequestHandle::const_iterator end = front.handles.end();
+            for (; it != end; ++it) {
+              bc_.cancelTransformableRequest(*it);
+            }
+
+            messageDropped(front.event, filter_failure_reasons::Unknown);
+
+            // Unlock the shared lock and get a unique lock. Upgradeable lock is used in transformable.
+            // There can only be one upgrade lock. It's important the cancelTransformableRequest not deadlock with transformable.
+            // They both require the transformable_requests_mutex_ in BufferCore.
+            // shared_lock.unlock();
+
+            // There is a very slight race condition if an older message arrives in this gap.
+            // std::unique_lock<std::mutex> unique_lock(messages_mutex_);
```

Comment:
I don't know the specifics of this for sure, but my guess is that this lock is there for a reason.  Do you know for sure that it is safe to remove it?

Reply:
Yeah, why it's unlock here is that the original use a shared clock of boost, but it needs to switch to use unique clock here, so it needs to unlock the previous shared clock to avoid dead clock. however, the current change has already used an unique clock and the unlock to change to use unique clock is not necessary now.